plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,388 @@
+	}
+	if xio.Password == "" {
+		return fmt.Errorf("Password cannot be empty")
+	}
+	if xio.URL == "" {
+		return fmt.Errorf("URL cannot be empty")
+	}
+	if len(xio.Collectors) == 0 {
+		xio.Collectors = []string{"bbus","clusters","ssds","volumes","xms"}
+	}
+
+	availableCollectors := []string{"bbus", "clusters", "ssds", "volumes", "xms"}
+	for _, collector := range xio.Collectors {
+		if !choice.Contains(collector,availableCollectors) {
+			return fmt.Errorf("Specified Collector Isn't Supported: " + collector)
+		}
+	}
+	
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	var err = xio.authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+
+			resp, err := xio.call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+
+			// Due to an inconsistency in the XtremIO API, the XMS endpoint
+			// returns a json array with XMSS as the result. Which is why this
+			// if statement here exists
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			}else{
+				arr = gjson.Get(resp, collector).Array()
+			}
+			
+			for _, item := range arr {
+				itemSplit := strings.Split(gjson.Get(item.Raw, "href").Str, "/")
+				url := ""
+				if len(itemSplit) < 1 {
+					continue
+				}
+				url = collector + "/" + itemSplit[len(itemSplit)-1]
+
+				switch collector {
+				case "bbus":
+					wg.Add(1)
+					go xio.gatherBBUs(acc, url, &wg)
+				case "clusters":
+					wg.Add(1)
+					go xio.gatherClusters(acc, url, &wg)
+				case "ssds":
+					wg.Add(1)
+					go xio.gatherSSDs(acc, url, &wg)
+				case "volumes":
+					wg.Add(1)
+					go xio.gatherVolumes(acc, url, &wg)
+				case "xms":
+					wg.Add(1)
+					go xio.gatherXMS(acc, url, &wg)
+				default:
+					acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				}
+			}
+		}(collector)
+	}
+	wg.Wait()
+
+	xio.Cookie = nil
+
+	return nil
+}
+
+func (xio *XtremIO) gatherBBUs(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+	}
+	tags := map[string]string{
+		"serial_number": gjson.Get(resp, "content.serial-number").Str,
+		"guid":          gjson.Get(resp, "content.guid").Str,
+		"power_feed":    gjson.Get(resp, "content.power-feed").Str,
+		"name":          gjson.Get(resp, "content.name").Str,
+		"model_name":    gjson.Get(resp, "content.model-name").Str,
+	}
+	fields := make(map[string]interface{})
+	fields["bbus_power"], err = strconv.Atoi(gjson.Get(resp, "content.power").Raw)
+	if err != nil {acc.AddError(err)}
```

Comment:
Please don't shortcut this
```suggestion
	if err != nil {
		acc.AddError(err)
	}
```
Do you mean to `return` here?

Reply:
I don't want to return here, just pass the error. An error retrieving one value doesn't necessarily mean the other values are irretrievable too.